grails-app/services/org/pih/warehouse/inventory/RequisitionStatusTransitionEventService.groovy

```diff
@@ -9,36 +9,36 @@
 **/
 package org.pih.warehouse.inventory
 
-import org.codehaus.groovy.grails.commons.GrailsApplication
-import org.pih.warehouse.core.MailService
 import org.pih.warehouse.core.Person
+import org.pih.warehouse.core.UserService
+import org.pih.warehouse.core.RoleType
+import org.pih.warehouse.report.NotificationService
 import org.pih.warehouse.requisition.Requisition
 import org.pih.warehouse.requisition.RequisitionService
 import org.pih.warehouse.requisition.RequisitionStatusTransitionEvent
+import org.pih.warehouse.requisition.RequisitionStatus
 import org.springframework.context.ApplicationEvent
 import org.springframework.context.ApplicationListener
 
 class RequisitionStatusTransitionEventService implements ApplicationListener<RequisitionStatusTransitionEvent> {
 
-    MailService mailService
-    GrailsApplication grailsApplication
     RequisitionService requisitionService
+    NotificationService notificationService
+    UserService userService
 
+    void onApplicationEvent(RequisitionStatusTransitionEvent event) {
+        if (event.newStatus == RequisitionStatus.PENDING_APPROVAL) {
+            Requisition req = event.requisition
 
-    void publishDefaultEmailNotifications(Requisition requisition, List<Person> receivers) {
-        def g = grailsApplication.mainContext.getBean('org.codehaus.groovy.grails.plugins.web.taglib.ApplicationTagLib')
-        String subject = "${requisition.requestNumber} ${requisition.name}"
-        String redirectToRequestsList = "${g.createLink(uri: "/stockMovement/list?direction=OUTBOUND&sourceType=ELECTRONIC", absolute: true)}"
-        GString body = "${g.render(template: "/email/approvalsAlert", model: [requisition: requisition, redirectUrl: redirectToRequestsList])}"
-
-        receivers.each {receiver ->
-            if (receiver.email) {
-                mailService.sendHtmlMail(subject, body, receiver.email)
+            List<Person> approvers
+            if (req.approvers?.size()) {
+                approvers = Person.findAllByIdInList(req.approvers.collect { it.id })
+            } else {
+                // if there are not assigned users as approvers then notify all approvers
+                approvers = userService.findUsersByRoleTypes(req.origin, [RoleType.ROLE_REQUISITION_APPROVER])
             }
+            notificationService.sendRequestPendingForApprovalNotification(req, approvers)
```

Comment:
Why can't we do all of the above in a method in the NotficationService? 

In fact let's create a method to resolve notification recipients (given a requisition). That way it's DRY and we we can extend it for other notifications. In addition, this makes it more testable. 
```
List<Person> resolveNotificationRecipients(Requisition requisition) { 
    List<Person> recipients = []
    if (requisition.status == RequisitionStatus.PENDING_APPROVAL) {
        if (requisition.approvers?.size()) {
            recipients = Person.findAllByIdInList(req.approvers.collect { it.id })
        } else {
            // if there are not assigned users as approvers then notify all approvers
            recipients = userService.findUsersByRoleTypes(req.origin, [RoleType.ROLE_REQUISITION_APPROVER])
        }
    }
    return recipients
}
```

Reply:
If it wasn't for the userService call I would have recommended moving the "recipients" logic to the Requisition domain since it has all of the information needed to query for that.

Reply:
Lastly, we should consider a slight renaming of the method as `sendRequestPendingForApprovalNotification(req, approvers)` is a little awkward.

I think `publishDefaultEmailNotifications` is closer to what I/we want although I'm glad you removed the "email" part since I'd like the notification service to be transport agnostic.

If we want the logic for what emails get sent in the event listener, then I guess
`publishRequisitionPendingApprovalNotifications` would be best.

If we want to allow the notification service to determine the what / when / who, then the following would be enough
```
publishNotifications(Requisition requisition)
```

Reply:
So in summary 
```
void onApplicationEvent(RequisitionStatusTransitionEvent event) {
        if (event.newStatus == RequisitionStatus.PENDING_APPROVAL) {
            notificationService.publishRequisitionPendingApprovalNotifications(event.requisition)
        } 
}
```

```
void publishRequisitionPendingApprovalNotifications(Requisition requisition) {
    List<Recipient> recipients = resolveNotificationRecipients(requisition)
    ...
}
```

Or 
```
void onApplicationEvent(RequisitionStatusTransitionEvent event) {
        notificationService.publishRequisitionStatusTransitionNotifications(event.requisition)
        requisitionService.triggerRequisitionStatusTransition(event.requisition, event.newStatus, event.createdBy)
}
```
with NotificationService looking like this 
```
void publishRequisitionStatusTransitionNotifications(Requisition requisition) {
    List<Recipient> recipients = resolveNotificationRecipients(requisition)
    if (requisition.status == RequisitionStatus.PENDING_APPROVAL) {
        notificationService.publishRequisitionPendingApprovalNotifications(event.requisition)
    } 
}

Reply:
I was in the middle of this comment when i had to stop so we can discuss this in more details on Monday's tech huddle.